

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Prerequisites &mdash; deepbiome 1.0.0+1.g7845235 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script src="_static/clipboard.min.js"></script>
        <script src="_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/copybutton.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Usage" href="usage.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> deepbiome
          

          
          </a>

          
            
            
              <div class="version">
                1.0.0+1.g7845235
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="DeepBiome_usage_classification.html">Getting start with the classification problem.</a></li>
<li class="toctree-l1"><a class="reference internal" href="DeepBiome_usage_regression.html">Getting start with the regression problem.</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Prerequisites</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#data-preprocessing">Data preprocessing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#with-a-list-of-k-inputs-for-k-times-repetition">With a list of <cite>k</cite> inputs for <cite>k</cite> times repetition</a></li>
<li class="toctree-l3"><a class="reference internal" href="#with-one-input-file-for-k-fold-cross-validation">With one input file for <cite>k</cite> fold cross-validation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#configuration">Configuration</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#preparing-the-configuration-about-the-network-information-network-info">Preparing the configuration about the network information (<cite>network_info</cite>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#network-info-model-info">network_info[‘model_info’]</a></li>
<li class="toctree-l4"><a class="reference internal" href="#network-info-architecture-info">network_info[‘architecture_info’]</a></li>
<li class="toctree-l4"><a class="reference internal" href="#network-info-training-info">network_info[‘training_info’]</a></li>
<li class="toctree-l4"><a class="reference internal" href="#network-info-validation-info">network_info[‘validation_info’]</a></li>
<li class="toctree-l4"><a class="reference internal" href="#network-info-test-info">network_info[‘test_info’]</a></li>
<li class="toctree-l4"><a class="reference internal" href="#example-for-the-network-info">Example for the <cite>network_info</cite></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#preparing-the-configuration-about-the-path-information-path-info">Preparing the configuration about the path information (<cite>path_info</cite>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#path-info-data-info">path_info[‘data_info’]</a></li>
<li class="toctree-l4"><a class="reference internal" href="#path-info-model-info">path_info[‘model_info’]</a></li>
<li class="toctree-l4"><a class="reference internal" href="#example-for-the-path-info-for-the-list-of-inputs">Example for the <cite>path_info</cite> for the list of inputs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#example-for-the-path-info-for-the-one-input-file">Example for the <cite>path_info</cite> for the one input file</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="example_with_one_input_file.html">Example : k fold cross-validation with an input file</a></li>
<li class="toctree-l1"><a class="reference internal" href="example_with_the_list_of_inputs.html">Example : k times repetition with the list of k input files</a></li>
<li class="toctree-l1"><a class="reference internal" href="release-history.html">Release History</a></li>
<li class="toctree-l1"><a class="reference internal" href="min_versions.html">Minimum Version of Python and NumPy</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">deepbiome</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Prerequisites</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/prerequisites.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="prerequisites">
<h1>Prerequisites<a class="headerlink" href="#prerequisites" title="Permalink to this headline">¶</a></h1>
<div class="section" id="data-preprocessing">
<h2>Data preprocessing<a class="headerlink" href="#data-preprocessing" title="Permalink to this headline">¶</a></h2>
<p>DeepBiome package takes microbiome abundance data as input and uses the phylogenetic taxonomy to guide the decision of the optimal number of layers and neurons in the deep learning architecture.</p>
<p>To use DeepBiome, you can experiment (1) <strong>`k` times repetition</strong> or (2) <strong>`k` fold cross-validation</strong>.
For each experiment, we asuume that the dataset is given by</p>
<blockquote>
<div><ol class="arabic simple">
<li><p><strong>A list of `k` input files for `k` times repetition.</strong>, or</p></li>
<li><p><strong>One input file for `k` fold cross-validation.</strong></p></li>
</ol>
</div></blockquote>
<div class="section" id="with-a-list-of-k-inputs-for-k-times-repetition">
<h3>With a list of <cite>k</cite> inputs for <cite>k</cite> times repetition<a class="headerlink" href="#with-a-list-of-k-inputs-for-k-times-repetition" title="Permalink to this headline">¶</a></h3>
<p>DeepBiome needs 4 data files as follows:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p><strong>the tree information</strong></p></li>
<li><p><strong>the list of the input files</strong> (each file has all sample’s information for one repetition)</p></li>
<li><p><strong>the list of the names of input files</strong></p></li>
<li><p><strong>y</strong></p></li>
</ol>
</div></blockquote>
<p>For <cite>k</cite> times repetition, we can use the list of <cite>k</cite> input files. Each file has all sample’s information for one repetition.
In addition, we can set <strong>the training index for each repetition</strong>. If we set the index file, DeepBiome builds the training set for each repetition based on each fold index in the index file. If not, DeepBiome will generate the index file locally.</p>
<p>Eath data should have the csv format as follows:</p>
<dl>
<dt>tree information (.csv)</dt><dd><p>A file about the phylogenetic tree information.
Below is an example file of the phylogenetic tree information dictionary</p>
<table class="longtable docutils align-default" id="id1">
<caption><span class="caption-text">Example of genus dictionary</span><a class="headerlink" href="#id1" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 17%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>Genus</p></td>
<td><p>Family</p></td>
<td><p>Order</p></td>
<td><p>Class</p></td>
<td><p>Phylum</p></td>
<td><p>Domain</p></td>
</tr>
<tr class="row-even"><td><p>Streptococcus</p></td>
<td><p>Streptococcaceae</p></td>
<td><p>Lactobacillales</p></td>
<td><p>Bacilli</p></td>
<td><p>Firmicutes</p></td>
<td><p>Bacteria</p></td>
</tr>
<tr class="row-odd"><td><p>Tropheryma</p></td>
<td><p>Cellulomonadaceae</p></td>
<td><p>Actinomycetales</p></td>
<td><p>Actinobacteria</p></td>
<td><p>Actinobacteria</p></td>
<td><p>Bacteria</p></td>
</tr>
<tr class="row-even"><td><p>Veillonella</p></td>
<td><p>Veillonellaceae</p></td>
<td><p>Selenomonadales</p></td>
<td><p>Negativicutes</p></td>
<td><p>Firmicutes</p></td>
<td><p>Bacteria</p></td>
</tr>
<tr class="row-odd"><td><p>Actinomyces</p></td>
<td><p>Actinomycetaceae</p></td>
<td><p>Actinomycetales</p></td>
<td><p>Actinobacteria</p></td>
<td><p>Actinobacteria</p></td>
<td><p>Bacteria</p></td>
</tr>
<tr class="row-even"><td><p>Flavobacterium</p></td>
<td><p>Flavobacteriaceae</p></td>
<td><p>Flavobacteriales</p></td>
<td><p>Flavobacteria</p></td>
<td><p>Bacteroidetes</p></td>
<td><p>Bacteria</p></td>
</tr>
<tr class="row-odd"><td><p>Prevotella</p></td>
<td><p>Prevotellaceae</p></td>
<td><p>Bacteroidales</p></td>
<td><p>Bacteroidia</p></td>
<td><p>Bacteroidetes</p></td>
<td><p>Bacteria</p></td>
</tr>
<tr class="row-even"><td><p>Porphyromonas</p></td>
<td><p>Porphyromonadaceae</p></td>
<td><p>Bacteroidales</p></td>
<td><p>Bacteroidia</p></td>
<td><p>Bacteroidetes</p></td>
<td><p>Bacteria</p></td>
</tr>
<tr class="row-odd"><td><p>Parvimonas</p></td>
<td><p>Clostridiales_Incertae_Sedis_XI</p></td>
<td><p>Clostridiales</p></td>
<td><p>Clostridia</p></td>
<td><p>Firmicutes</p></td>
<td><p>Bacteria</p></td>
</tr>
</tbody>
</table>
</dd>
<dt>list of the names of <cite>k</cite> input files (.csv)</dt><dd><p>If we want to use the list of the input files, we need the make a list of the names of each input file.
Below is an example file for <cite>k=4</cite> repetition.</p>
<table class="longtable docutils align-default" id="id2">
<caption><span class="caption-text">Example of the list of <cite>k</cite> input file names</span><a class="headerlink" href="#id2" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 100%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>gcount_0001.csv</p></td>
</tr>
<tr class="row-even"><td><p>gcount_0002.csv</p></td>
</tr>
<tr class="row-odd"><td><p>gcount_0003.csv</p></td>
</tr>
<tr class="row-even"><td><p>gcount_0004.csv</p></td>
</tr>
</tbody>
</table>
</dd>
<dt>list of <cite>k</cite> input files</dt><dd><p>Each file should have each repetition’s sample microbiome abandunce.
Below is an example file for <cite>k=4</cite> repetition. This example is <cite>gcount_0001.csv</cite> for the first repetition in the list of the names of input files above. This file has the 4 samples’ microbiome abandunce.</p>
<table class="longtable docutils align-default" id="id3">
<caption><span class="caption-text">Example of one input file (.csv) of <cite>k</cite> inputs.</span><a class="headerlink" href="#id3" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>Streptococcus</p></td>
<td><p>Tropheryma</p></td>
<td><p>Veillonella</p></td>
<td><p>Actinomyces</p></td>
<td><p>Flavobacterium</p></td>
<td><p>Prevotella</p></td>
<td><p>Porphyromonas</p></td>
<td><p>Parvimonas</p></td>
<td><p>Fusobacterium</p></td>
<td><p>Propionibacterium</p></td>
<td><p>Gemella</p></td>
<td><p>Rothia</p></td>
<td><p>Granulicatella</p></td>
<td><p>Neisseria</p></td>
<td><p>Lactobacillus</p></td>
<td><p>Megasphaera</p></td>
<td><p>Catonella</p></td>
<td><p>Atopobium</p></td>
<td><p>Campylobacter</p></td>
<td><p>Capnocytophaga</p></td>
<td><p>Solobacterium</p></td>
<td><p>Moryella</p></td>
<td><p>TM7_genera_incertae_sedis</p></td>
<td><p>Staphylococcus</p></td>
<td><p>Filifactor</p></td>
<td><p>Oribacterium</p></td>
<td><p>Burkholderia</p></td>
<td><p>Sneathia</p></td>
<td><p>Treponema</p></td>
<td><p>Moraxella</p></td>
<td><p>Haemophilus</p></td>
<td><p>Selenomonas</p></td>
<td><p>Corynebacterium</p></td>
<td><p>Rhizobium</p></td>
<td><p>Bradyrhizobium</p></td>
<td><p>Methylobacterium</p></td>
<td><p>OD1_genera_incertae_sedis</p></td>
<td><p>Finegoldia</p></td>
<td><p>Microbacterium</p></td>
<td><p>Sphingomonas</p></td>
<td><p>Chryseobacterium</p></td>
<td><p>Bacteroides</p></td>
<td><p>Bdellovibrio</p></td>
<td><p>Streptophyta</p></td>
<td><p>Lachnospiracea_incertae_sedis</p></td>
<td><p>Paracoccus</p></td>
<td><p>Fastidiosipila</p></td>
<td><p>Pseudonocardia</p></td>
</tr>
<tr class="row-even"><td><p>841</p></td>
<td><p>0</p></td>
<td><p>813</p></td>
<td><p>505</p></td>
<td><p>5</p></td>
<td><p>3224</p></td>
<td><p>0</p></td>
<td><p>362</p></td>
<td><p>11</p></td>
<td><p>65</p></td>
<td><p>156</p></td>
<td><p>1</p></td>
<td><p>55</p></td>
<td><p>0</p></td>
<td><p>1</p></td>
<td><p>20</p></td>
<td><p>382</p></td>
<td><p>1</p></td>
<td><p>333</p></td>
<td><p>24</p></td>
<td><p>80</p></td>
<td><p>43</p></td>
<td><p>309</p></td>
<td><p>2</p></td>
<td><p>3</p></td>
<td><p>4</p></td>
<td><p>0</p></td>
<td><p>1</p></td>
<td><p>32</p></td>
<td><p>0</p></td>
<td><p>2</p></td>
<td><p>4</p></td>
<td><p>382</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>96</p></td>
<td><p>23</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>87</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>2133</p></td>
</tr>
<tr class="row-odd"><td><p>1445</p></td>
<td><p>0</p></td>
<td><p>1</p></td>
<td><p>573</p></td>
<td><p>0</p></td>
<td><p>1278</p></td>
<td><p>82</p></td>
<td><p>85</p></td>
<td><p>69</p></td>
<td><p>154</p></td>
<td><p>436</p></td>
<td><p>3</p></td>
<td><p>0</p></td>
<td><p>61</p></td>
<td><p>440</p></td>
<td><p>0</p></td>
<td><p>394</p></td>
<td><p>83</p></td>
<td><p>33</p></td>
<td><p>123</p></td>
<td><p>0</p></td>
<td><p>49</p></td>
<td><p>414</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>37</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>42</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>384</p></td>
<td><p>27</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>146</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>1</p></td>
<td><p>2</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>3638</p></td>
</tr>
<tr class="row-even"><td><p>1259</p></td>
<td><p>0</p></td>
<td><p>805</p></td>
<td><p>650</p></td>
<td><p>0</p></td>
<td><p>1088</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>74</p></td>
<td><p>0</p></td>
<td><p>155</p></td>
<td><p>228</p></td>
<td><p>430</p></td>
<td><p>765</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>11</p></td>
<td><p>102</p></td>
<td><p>68</p></td>
<td><p>90</p></td>
<td><p>77</p></td>
<td><p>83</p></td>
<td><p>322</p></td>
<td><p>10</p></td>
<td><p>0</p></td>
<td><p>7</p></td>
<td><p>0</p></td>
<td><p>122</p></td>
<td><p>76</p></td>
<td><p>0</p></td>
<td><p>1</p></td>
<td><p>25</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>44</p></td>
<td><p>13</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>2</p></td>
<td><p>8</p></td>
<td><p>1</p></td>
<td><p>39</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>3445</p></td>
</tr>
<tr class="row-odd"><td><p>982</p></td>
<td><p>0</p></td>
<td><p>327</p></td>
<td><p>594</p></td>
<td><p>0</p></td>
<td><p>960</p></td>
<td><p>81</p></td>
<td><p>19</p></td>
<td><p>9</p></td>
<td><p>0</p></td>
<td><p>45</p></td>
<td><p>457</p></td>
<td><p>1049</p></td>
<td><p>0</p></td>
<td><p>3</p></td>
<td><p>450</p></td>
<td><p>19</p></td>
<td><p>170</p></td>
<td><p>388</p></td>
<td><p>147</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>41</p></td>
<td><p>63</p></td>
<td><p>0</p></td>
<td><p>1</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>121</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>1</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>344</p></td>
<td><p>0</p></td>
<td><p>157</p></td>
<td><p>1</p></td>
<td><p>0</p></td>
<td><p>4</p></td>
<td><p>60</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>3507</p></td>
</tr>
</tbody>
</table>
</dd>
<dt>y (.csv)</dt><dd><p>One column contains y samples for one repetition.
Below is an example file for <cite>k=4</cite> repetition. For each repetition (column) has outputs of 4 samples for each repeatition.</p>
<table class="longtable docutils align-default" id="id4">
<caption><span class="caption-text">Example of y file (.csv)</span><a class="headerlink" href="#id4" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>V1</p></td>
<td><p>V2</p></td>
<td><p>V3</p></td>
<td><p>V4</p></td>
</tr>
<tr class="row-even"><td><p>1.0</p></td>
<td><p>0.0</p></td>
<td><p>1.0</p></td>
<td><p>1.0</p></td>
</tr>
<tr class="row-odd"><td><p>1.0</p></td>
<td><p>1.0</p></td>
<td><p>1.0</p></td>
<td><p>1.0</p></td>
</tr>
<tr class="row-even"><td><p>0.0</p></td>
<td><p>1.0</p></td>
<td><p>0.0</p></td>
<td><p>1.0</p></td>
</tr>
<tr class="row-odd"><td><p>0.0</p></td>
<td><p>1.0</p></td>
<td><p>1.0</p></td>
<td><p>0.0</p></td>
</tr>
</tbody>
</table>
</dd>
<dt>index for training set for each repetition (.csv)</dt><dd><p>For each repetition, we have to set the training and test set. If the index file is given, DeepBiome sets the training set and test set based on the index file. Below is the example of the index file. Each column has the training indices for each repetition. DeepBiome will only use the samples in this index set for training. Below is an example for <cite>k=4</cite> repetition</p>
<table class="longtable docutils align-default" id="id5">
<caption><span class="caption-text">Example of index file (.csv)</span><a class="headerlink" href="#id5" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>V1</p></td>
<td><p>V2</p></td>
<td><p>V3</p></td>
<td><p>V4</p></td>
</tr>
<tr class="row-even"><td><p>0</p></td>
<td><p>1</p></td>
<td><p>2</p></td>
<td><p>3</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>2</p></td>
<td><p>3</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-even"><td><p>2</p></td>
<td><p>3</p></td>
<td><p>0</p></td>
<td><p>1</p></td>
</tr>
</tbody>
</table>
</dd>
</dl>
<p>In the example above, we used the first 3 rows of the first column in <cite>y.csv</cite> for the training set in the first repetition.</p>
</div>
<div class="section" id="with-one-input-file-for-k-fold-cross-validation">
<h3>With one input file for <cite>k</cite> fold cross-validation<a class="headerlink" href="#with-one-input-file-for-k-fold-cross-validation" title="Permalink to this headline">¶</a></h3>
<p>DeepBiome needs 3 data files as follows:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p><strong>the tree information</strong></p></li>
<li><p><strong>the input file</strong></p></li>
<li><p><strong>y</strong></p></li>
</ol>
</div></blockquote>
<p>For <cite>k</cite> fold cross-validation, we can use an input file.
In addition, we can set <strong>the training index for each fold</strong>. If we set the index file, DeepBiome builds the training set for each fold based on each fold index in the index file. If not, DeepBiome will generate the index file locally.</p>
<p>Eath data should have the csv format as follows:</p>
<dl>
<dt>tree information (.csv)</dt><dd><p>A file about the phylogenetic tree information.
Below is an example file of the phylogenetic tree information dictionary</p>
<table class="longtable docutils align-default" id="id6">
<caption><span class="caption-text">Example of genus dictionary</span><a class="headerlink" href="#id6" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 17%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>Genus</p></td>
<td><p>Family</p></td>
<td><p>Order</p></td>
<td><p>Class</p></td>
<td><p>Phylum</p></td>
<td><p>Domain</p></td>
</tr>
<tr class="row-even"><td><p>Streptococcus</p></td>
<td><p>Streptococcaceae</p></td>
<td><p>Lactobacillales</p></td>
<td><p>Bacilli</p></td>
<td><p>Firmicutes</p></td>
<td><p>Bacteria</p></td>
</tr>
<tr class="row-odd"><td><p>Tropheryma</p></td>
<td><p>Cellulomonadaceae</p></td>
<td><p>Actinomycetales</p></td>
<td><p>Actinobacteria</p></td>
<td><p>Actinobacteria</p></td>
<td><p>Bacteria</p></td>
</tr>
<tr class="row-even"><td><p>Veillonella</p></td>
<td><p>Veillonellaceae</p></td>
<td><p>Selenomonadales</p></td>
<td><p>Negativicutes</p></td>
<td><p>Firmicutes</p></td>
<td><p>Bacteria</p></td>
</tr>
<tr class="row-odd"><td><p>Actinomyces</p></td>
<td><p>Actinomycetaceae</p></td>
<td><p>Actinomycetales</p></td>
<td><p>Actinobacteria</p></td>
<td><p>Actinobacteria</p></td>
<td><p>Bacteria</p></td>
</tr>
<tr class="row-even"><td><p>Flavobacterium</p></td>
<td><p>Flavobacteriaceae</p></td>
<td><p>Flavobacteriales</p></td>
<td><p>Flavobacteria</p></td>
<td><p>Bacteroidetes</p></td>
<td><p>Bacteria</p></td>
</tr>
<tr class="row-odd"><td><p>Prevotella</p></td>
<td><p>Prevotellaceae</p></td>
<td><p>Bacteroidales</p></td>
<td><p>Bacteroidia</p></td>
<td><p>Bacteroidetes</p></td>
<td><p>Bacteria</p></td>
</tr>
<tr class="row-even"><td><p>Porphyromonas</p></td>
<td><p>Porphyromonadaceae</p></td>
<td><p>Bacteroidales</p></td>
<td><p>Bacteroidia</p></td>
<td><p>Bacteroidetes</p></td>
<td><p>Bacteria</p></td>
</tr>
<tr class="row-odd"><td><p>Parvimonas</p></td>
<td><p>Clostridiales_Incertae_Sedis_XI</p></td>
<td><p>Clostridiales</p></td>
<td><p>Clostridia</p></td>
<td><p>Firmicutes</p></td>
<td><p>Bacteria</p></td>
</tr>
</tbody>
</table>
</dd>
<dt>input file</dt><dd><p>Input file has the microbiome abandunce of each sample.
Below is an example file with the 4 samples’ microbiome abandunce.</p>
<table class="longtable docutils align-default" id="id7">
<caption><span class="caption-text">Example of input file (.csv)</span><a class="headerlink" href="#id7" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
<col style="width: 2%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>Streptococcus</p></td>
<td><p>Tropheryma</p></td>
<td><p>Veillonella</p></td>
<td><p>Actinomyces</p></td>
<td><p>Flavobacterium</p></td>
<td><p>Prevotella</p></td>
<td><p>Porphyromonas</p></td>
<td><p>Parvimonas</p></td>
<td><p>Fusobacterium</p></td>
<td><p>Propionibacterium</p></td>
<td><p>Gemella</p></td>
<td><p>Rothia</p></td>
<td><p>Granulicatella</p></td>
<td><p>Neisseria</p></td>
<td><p>Lactobacillus</p></td>
<td><p>Megasphaera</p></td>
<td><p>Catonella</p></td>
<td><p>Atopobium</p></td>
<td><p>Campylobacter</p></td>
<td><p>Capnocytophaga</p></td>
<td><p>Solobacterium</p></td>
<td><p>Moryella</p></td>
<td><p>TM7_genera_incertae_sedis</p></td>
<td><p>Staphylococcus</p></td>
<td><p>Filifactor</p></td>
<td><p>Oribacterium</p></td>
<td><p>Burkholderia</p></td>
<td><p>Sneathia</p></td>
<td><p>Treponema</p></td>
<td><p>Moraxella</p></td>
<td><p>Haemophilus</p></td>
<td><p>Selenomonas</p></td>
<td><p>Corynebacterium</p></td>
<td><p>Rhizobium</p></td>
<td><p>Bradyrhizobium</p></td>
<td><p>Methylobacterium</p></td>
<td><p>OD1_genera_incertae_sedis</p></td>
<td><p>Microbacterium</p></td>
<td><p>Sphingomonas</p></td>
<td><p>Chryseobacterium</p></td>
<td><p>Bdellovibrio</p></td>
<td><p>Streptophyta</p></td>
<td><p>Finegoldia</p></td>
<td><p>Bacteroides</p></td>
<td><p>Lachnospiracea_incertae_sedis</p></td>
<td><p>Paracoccus</p></td>
<td><p>Fastidiosipila</p></td>
<td><p>Pseudonocardia</p></td>
</tr>
<tr class="row-even"><td><p>6244.0</p></td>
<td><p>0.0</p></td>
<td><p>2985.0</p></td>
<td><p>204.0</p></td>
<td><p>5.0</p></td>
<td><p>3548.0</p></td>
<td><p>308.0</p></td>
<td><p>53.0</p></td>
<td><p>506.0</p></td>
<td><p>3.0</p></td>
<td><p>324.0</p></td>
<td><p>669.0</p></td>
<td><p>795.0</p></td>
<td><p>0.0</p></td>
<td><p>3686.0</p></td>
<td><p>6.0</p></td>
<td><p>41.0</p></td>
<td><p>609.0</p></td>
<td><p>11.0</p></td>
<td><p>0.0</p></td>
<td><p>3.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>31.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>3.0</p></td>
<td><p>0.0</p></td>
<td><p>93.0</p></td>
<td><p>0.0</p></td>
<td><p>1.0</p></td>
<td><p>9.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>3.0</p></td>
<td><p>2.0</p></td>
<td><p>5.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>2.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
</tr>
<tr class="row-odd"><td><p>3573.0</p></td>
<td><p>3.0</p></td>
<td><p>2566.0</p></td>
<td><p>975.0</p></td>
<td><p>0.0</p></td>
<td><p>7195.0</p></td>
<td><p>377.0</p></td>
<td><p>39.0</p></td>
<td><p>442.0</p></td>
<td><p>69.0</p></td>
<td><p>602.0</p></td>
<td><p>45.0</p></td>
<td><p>527.0</p></td>
<td><p>4536.0</p></td>
<td><p>324.0</p></td>
<td><p>105.0</p></td>
<td><p>626.0</p></td>
<td><p>400.0</p></td>
<td><p>1130.0</p></td>
<td><p>1132.0</p></td>
<td><p>127.0</p></td>
<td><p>9.0</p></td>
<td><p>426.0</p></td>
<td><p>16.0</p></td>
<td><p>0.0</p></td>
<td><p>150.0</p></td>
<td><p>1.0</p></td>
<td><p>244.0</p></td>
<td><p>248.0</p></td>
<td><p>0.0</p></td>
<td><p>19.0</p></td>
<td><p>26.0</p></td>
<td><p>9.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>14.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
</tr>
<tr class="row-even"><td><p>516.0</p></td>
<td><p>0.0</p></td>
<td><p>227.0</p></td>
<td><p>39.0</p></td>
<td><p>0.0</p></td>
<td><p>180.0</p></td>
<td><p>54.0</p></td>
<td><p>38.0</p></td>
<td><p>80.0</p></td>
<td><p>2.0</p></td>
<td><p>84.0</p></td>
<td><p>26.0</p></td>
<td><p>18.0</p></td>
<td><p>0.0</p></td>
<td><p>2235.0</p></td>
<td><p>4.0</p></td>
<td><p>3.0</p></td>
<td><p>3.0</p></td>
<td><p>8.0</p></td>
<td><p>2.0</p></td>
<td><p>10.0</p></td>
<td><p>2.0</p></td>
<td><p>7.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>78.0</p></td>
<td><p>7.0</p></td>
<td><p>0.0</p></td>
<td><p>4.0</p></td>
<td><p>6.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>1.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
</tr>
<tr class="row-odd"><td><p>881.0</p></td>
<td><p>1.0</p></td>
<td><p>47.0</p></td>
<td><p>4.0</p></td>
<td><p>0.0</p></td>
<td><p>13.0</p></td>
<td><p>5.0</p></td>
<td><p>1.0</p></td>
<td><p>108.0</p></td>
<td><p>1.0</p></td>
<td><p>7.0</p></td>
<td><p>6.0</p></td>
<td><p>6.0</p></td>
<td><p>117.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>2.0</p></td>
<td><p>0.0</p></td>
<td><p>4.0</p></td>
<td><p>6.0</p></td>
<td><p>0.0</p></td>
<td><p>5.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>9.0</p></td>
<td><p>3.0</p></td>
<td><p>6.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>2.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>1.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
</tr>
</tbody>
</table>
</dd>
<dt>y (.csv)</dt><dd><p>Below is an example file of the outputs of 4 samples.</p>
<table class="longtable docutils align-default" id="id8">
<caption><span class="caption-text">Example of y file (.csv)</span><a class="headerlink" href="#id8" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 100%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>y</p></td>
</tr>
<tr class="row-even"><td><p>0.78</p></td>
</tr>
<tr class="row-odd"><td><p>1.01</p></td>
</tr>
<tr class="row-even"><td><p>0.92</p></td>
</tr>
<tr class="row-odd"><td><p>0.91</p></td>
</tr>
</tbody>
</table>
</dd>
<dt>index for training set for each fold (.csv)</dt><dd><p>For each fold, we have to set the training and test set. If the index file is given, DeepBiome sets the training set and test set based on the index file. Below is an example of the index file. Each column has the training indices for each fold. DeepBiome will only use the samples in this index set for training. Below is an example for <cite>k=4</cite> fold</p>
<table class="longtable docutils align-default" id="id9">
<caption><span class="caption-text">Example of index file (.csv)</span><a class="headerlink" href="#id9" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>V1</p></td>
<td><p>V2</p></td>
<td><p>V3</p></td>
<td><p>V4</p></td>
</tr>
<tr class="row-even"><td><p>0</p></td>
<td><p>1</p></td>
<td><p>2</p></td>
<td><p>3</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>2</p></td>
<td><p>3</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-even"><td><p>2</p></td>
<td><p>3</p></td>
<td><p>0</p></td>
<td><p>1</p></td>
</tr>
</tbody>
</table>
</dd>
</dl>
<p>In the example above, we used the first 3 rows of the first column in <cite>y.csv</cite> for the training set in the first fold.</p>
</div>
</div>
<div class="section" id="configuration">
<h2>Configuration<a class="headerlink" href="#configuration" title="Permalink to this headline">¶</a></h2>
<p>For detailed configuration, we used python dictionary as inputs for the main training function.</p>
<div class="section" id="preparing-the-configuration-about-the-network-information-network-info">
<h3>Preparing the configuration about the network information (<cite>network_info</cite>)<a class="headerlink" href="#preparing-the-configuration-about-the-network-information-network-info" title="Permalink to this headline">¶</a></h3>
<p>To give the information about the training hyper-parameter, we provide a dictionary of configuration to the <cite>netowrk_info</cite> field. Alternatively  we can use the configufation file (.cfg).</p>
<p>Configuration for the network training should include the information about:</p>
<dl class="field-list simple">
<dt class="field-odd">model_info</dt>
<dd class="field-odd"><p>about the training method and metrics</p>
</dd>
<dt class="field-even">architecture_info</dt>
<dd class="field-even"><p>about the architecture options</p>
</dd>
<dt class="field-odd">training_info</dt>
<dd class="field-odd"><p>about the hyper-parameter for training (not required for testing and prediction)</p>
</dd>
<dt class="field-even">validation_info</dt>
<dd class="field-even"><p>about the hyper-parameter for validation (not required for testing and prediction)</p>
</dd>
<dt class="field-odd">test_info</dt>
<dd class="field-odd"><p>about the hyper-parameter for testing</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You don’t have to fill the options if it has a default value.</p>
</div>
<div class="section" id="network-info-model-info">
<h4>network_info[‘model_info’]<a class="headerlink" href="#network-info-model-info" title="Permalink to this headline">¶</a></h4>
<p>Detailed options for the <cite>model_info</cite> field are as follows.</p>
<dl class="field-list">
<dt class="field-odd">network_clas</dt>
<dd class="field-odd"><p>DeepBiome network class (default=’DeepBiomeNetwork’).</p>
</dd>
<dt class="field-even">reader_class</dt>
<dd class="field-even"><p>reader classes</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 30%" />
<col style="width: 70%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>possible options</p></th>
<th class="head"><p>explanation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>“MicroBiomeRegressionReader”</p></td>
<td><p>Microbiome adandunce data reader for regression problem</p></td>
</tr>
<tr class="row-odd"><td><p>“MicroBiomeClassificationReader”</p></td>
<td><p>Microbiome adandunce data reader for classification problem</p></td>
</tr>
</tbody>
</table>
</dd>
<dt class="field-odd">optimizer</dt>
<dd class="field-odd"><p>optimization methods for training the network. We used the optimizers implemented in Keras (See <a class="reference external" href="https://keras.io/optimizers/">Optimizer</a>).</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 20%" />
<col style="width: 80%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>possible options</p></th>
<th class="head"><p>explanation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>“adam”</p></td>
<td><p>Adam optimizer</p></td>
</tr>
<tr class="row-odd"><td><p>“sgd”</p></td>
<td><p>stocastic gradient decent optimizer</p></td>
</tr>
</tbody>
</table>
</dd>
<dt class="field-even">lr</dt>
<dd class="field-even"><p>learning rate for the optimizor. (float between 0 ~ 1)</p>
</dd>
<dt class="field-odd">decay</dt>
<dd class="field-odd"><p>learning late decay ratio for the optimizer. (float between 0 ~ 1)</p>
</dd>
<dt class="field-even">loss</dt>
<dd class="field-even"><p>loss functions for training the network</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 26%" />
<col style="width: 74%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>possible options</p></th>
<th class="head"><p>explanation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>“mean_squared_error”</p></td>
<td><p>for regression problem</p></td>
</tr>
<tr class="row-odd"><td><p>“binary_crossentropy”</p></td>
<td><p>for binary classification problem</p></td>
</tr>
<tr class="row-even"><td><p>“categorical_crossentropy”</p></td>
<td><p>for multi-class classification problem</p></td>
</tr>
</tbody>
</table>
</dd>
<dt class="field-odd">metrics</dt>
<dd class="field-odd"><p>additional metrics to check the model performance</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 26%" />
<col style="width: 74%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>possible options</p></th>
<th class="head"><p>explanation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>“correlation_coefficient”</p></td>
<td><p>Pearson correlation coefficient (-1 ~ 1)</p></td>
</tr>
<tr class="row-odd"><td><p>“binary_accuracy”</p></td>
<td><p>Accuracy for binary classification problem (0 ~ 1)</p></td>
</tr>
<tr class="row-even"><td><p>“categorical_accuracy”</p></td>
<td><p>Accuracy for multi-class classification problem (0 ~ 1)</p></td>
</tr>
<tr class="row-odd"><td><p>“sensitivity”</p></td>
<td><p>Sensitivity (0 ~ 1)</p></td>
</tr>
<tr class="row-even"><td><p>“specificity”</p></td>
<td><p>Specificity (0 ~ 1)</p></td>
</tr>
<tr class="row-odd"><td><p>“gmeasure”</p></td>
<td><p>(Sensitivity * Specificity) ^ (0.5) (0 ~ 1)</p></td>
</tr>
<tr class="row-even"><td><p>“auc”</p></td>
<td><p>Area under the receiver operating characteristics (0 ~ 1)</p></td>
</tr>
<tr class="row-odd"><td><p>“precision”</p></td>
<td><p>Precision (0 ~ 1)</p></td>
</tr>
<tr class="row-even"><td><p>“recall”</p></td>
<td><p>Recall (0 ~ 1)</p></td>
</tr>
<tr class="row-odd"><td><p>“f1”</p></td>
<td><p>F1 score (0 ~ 1)</p></td>
</tr>
</tbody>
</table>
</dd>
<dt class="field-even">taxa_selection_metrics</dt>
<dd class="field-even"><p>metrics for the texa selection performance</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 26%" />
<col style="width: 74%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>possible options</p></th>
<th class="head"><p>explanation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>“accuracy”</p></td>
<td><p>Accuracy (-1 ~ 1)</p></td>
</tr>
<tr class="row-odd"><td><p>“sensitivity”</p></td>
<td><p>Sensitivity (0 ~ 1)</p></td>
</tr>
<tr class="row-even"><td><p>“specificity”</p></td>
<td><p>Specificity (0 ~ 1)</p></td>
</tr>
<tr class="row-odd"><td><p>“gmeasure”</p></td>
<td><p>(Sensitivity * Specificity) ^ (0.5) (0 ~ 1)</p></td>
</tr>
</tbody>
</table>
</dd>
<dt class="field-odd">normalizer</dt>
<dd class="field-odd"><p>normalizer for the input data (default=`normalize_minmax`)</p>
</dd>
</dl>
</div>
<div class="section" id="network-info-architecture-info">
<h4>network_info[‘architecture_info’]<a class="headerlink" href="#network-info-architecture-info" title="Permalink to this headline">¶</a></h4>
<p>Detailed options for the <cite>architecture_info</cite> field are as follows.</p>
<p>Combination of the options below will provide you the network training method <cite>DNN</cite>, <cite>DNN+L1</cite> and <cite>Deepbiome</cite> in the reference (url. TBD)</p>
<dl class="field-list">
<dt class="field-odd">weight_initial</dt>
<dd class="field-odd"><p>network weight initialization</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 75%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>possible options</p></th>
<th class="head"><p>explanation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>“glorot_uniform”</p></td>
<td><p>Glorot uniform initializer (defualt)</p></td>
</tr>
<tr class="row-odd"><td><p>“he_normal”</p></td>
<td><p>He normal initializer</p></td>
</tr>
<tr class="row-even"><td><p>“phylogenetic_tree”</p></td>
<td><p>weight within the tree connection: 1; weight without the tree connection: 0</p></td>
</tr>
<tr class="row-odd"><td><p>“phylogenetic_tree_glorot_uniform”</p></td>
<td><p>weight within the tree connection: <cite>glorot_uniform</cite>; weight without the tree connection: 0</p></td>
</tr>
<tr class="row-even"><td><p>“phylogenetic_tree_he_normal”</p></td>
<td><p>weight within the tree connection: <cite>he_normal</cite>; weight without the tree connection: 0</p></td>
</tr>
</tbody>
</table>
</dd>
<dt class="field-even">weight_l1_penalty</dt>
<dd class="field-even"><p><span class="math notranslate nohighlight">\(\lambda\)</span> for l1 penalty (float. defaut = 0)</p>
</dd>
<dt class="field-odd">weight_l2_penalty</dt>
<dd class="field-odd"><p><span class="math notranslate nohighlight">\(\lambda\)</span> for l2 penalty (float. defaut = 0)</p>
</dd>
<dt class="field-even">weight_decay</dt>
<dd class="field-even"><p><strong>DeepBiome with the phylogenetic tree based weight decay method</strong> (default = “”: without deepbiome weight decay method)</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 23%" />
<col style="width: 77%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>possible options</p></th>
<th class="head"><p>explanation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>“phylogenetic_tree”</p></td>
<td><p>weight decay method based on the phylogenetic tree information with small amout of noise (<span class="math notranslate nohighlight">\(\epsilon \le 1e-2\)</span>)</p></td>
</tr>
<tr class="row-odd"><td><p>“phylogenetic_tree_wo_noise”</p></td>
<td><p>weight decay method based on the phylogenetic tree information without any noise outside the tree</p></td>
</tr>
</tbody>
</table>
</dd>
<dt class="field-odd">batch_normalization</dt>
<dd class="field-odd"><p>options for adding the batch normalization for each convolutional layer (default = <cite>False</cite>)</p>
</dd>
<dt class="field-even">drop_out</dt>
<dd class="field-even"><p>options for adding the drop out for each convolutional layer with given ratio (default = 0)</p>
</dd>
</dl>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>Example of the combination of the options in the reference paper (url TBD):</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 24%" />
<col style="width: 76%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>training method</p></th>
<th class="head"><p>combination of the options</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>DNN</p></td>
<td><p>“weight_initial”=”glorot_uniform”</p></td>
</tr>
<tr class="row-odd"><td><p>DNN+L1</p></td>
<td><p>“weight_initial”=”glorot_uniform”, “weight_l1_penalty”=”0.01”</p></td>
</tr>
<tr class="row-even"><td><p>DeepBiome</p></td>
<td><p>“weight_initial”=”glorot_uniform”, “weight_deacy”=”phylogenetic_tree”</p></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="network-info-training-info">
<h4>network_info[‘training_info’]<a class="headerlink" href="#network-info-training-info" title="Permalink to this headline">¶</a></h4>
<p>Detailed options for the <cite>training_info</cite> field are as follows.</p>
<dl class="field-list">
<dt class="field-odd">epochs</dt>
<dd class="field-odd"><p>number of the epoch for training (integer)</p>
</dd>
<dt class="field-even">batch_size</dt>
<dd class="field-even"><p>number of the batch size for each mini-batch (integer)</p>
</dd>
<dt class="field-odd">callbacks</dt>
<dd class="field-odd"><p>callback class implemented in Keras (See <a class="reference external" href="https://keras.io/callbacks/">Callbacks</a>)</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 20%" />
<col style="width: 80%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>possible options</p></th>
<th class="head"><p>explanation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>“ModelCheckpoint”</p></td>
<td><p>save the best model weight based on the monitor (See <a class="reference external" href="https://keras.io/callbacks/#modelcheckpoint">ModelCheckpoint</a>)</p></td>
</tr>
<tr class="row-odd"><td><p>“EarlyStopping”</p></td>
<td><p>early stopping the training before the number of epochs <cite>epochs</cite> based on the monitor (See <a class="reference external" href="https://keras.io/callbacks/#earlystopping">EarlyStopping</a>)</p></td>
</tr>
</tbody>
</table>
</dd>
<dt class="field-even">monitor</dt>
<dd class="field-even"><p>monitor value for the <cite>ModelCheckpoint</cite>, <cite>EarlyStoppoing</cite> callbacks (e.g.  <cite>val_loss</cite>, <cite>val_accuray</cite>)</p>
</dd>
<dt class="field-odd">mode</dt>
<dd class="field-odd"><p>how to use the monitor value for the <cite>ModelCheckpoint</cite>, <cite>EarlyStopping</cite> callbacks</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 26%" />
<col style="width: 74%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>possible options</p></th>
<th class="head"><p>explanation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>“min”</p></td>
<td><p>for example: when using the monitor <cite>val_loss</cite></p></td>
</tr>
<tr class="row-odd"><td><p>“max”</p></td>
<td><p>for example: when using the monitor <cite>val_accuray</cite></p></td>
</tr>
</tbody>
</table>
</dd>
<dt class="field-even">patience</dt>
<dd class="field-even"><p>patience for the EarlyStopping callback (integer; default = 20)</p>
</dd>
<dt class="field-odd">min_delta</dt>
<dd class="field-odd"><p>the minimum threshold for the ModelCheckpoint, EarlyStopping callbacks (float; default = 1e-4)</p>
</dd>
</dl>
</div>
<div class="section" id="network-info-validation-info">
<h4>network_info[‘validation_info’]<a class="headerlink" href="#network-info-validation-info" title="Permalink to this headline">¶</a></h4>
<p>Detailed options for the <cite>validation_info</cite> field are as follows.</p>
<dl class="field-list simple">
<dt class="field-odd">validation_size</dt>
<dd class="field-odd"><p>the ratio of the number of the samples in the validation set / the number of the samples in the training set(e.g. “0.2”)</p>
</dd>
<dt class="field-even">batch_size</dt>
<dd class="field-even"><p>the batch size for each mini-batch. If “None”, use the whole number of the sample as one mini-batch. (defualt = “None”)</p>
</dd>
</dl>
</div>
<div class="section" id="network-info-test-info">
<h4>network_info[‘test_info’]<a class="headerlink" href="#network-info-test-info" title="Permalink to this headline">¶</a></h4>
<p>Detailed options for the <cite>test_info</cite> field are as follows.</p>
<dl class="field-list simple">
<dt class="field-odd">batch_size</dt>
<dd class="field-odd"><p>the batch size for each mini-batch. If “None”, use the whole number of the sample as one mini-batch. (defualt = “None”)</p>
</dd>
</dl>
</div>
<div class="section" id="example-for-the-network-info">
<h4>Example for the <cite>network_info</cite><a class="headerlink" href="#example-for-the-network-info" title="Permalink to this headline">¶</a></h4>
<p>This is the example of the configuration dictionary: <cite>network_info</cite> dictionary</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">network_info</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;architecture_info&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;batch_normalization&#39;</span><span class="p">:</span> <span class="s1">&#39;False&#39;</span><span class="p">,</span>
        <span class="s1">&#39;drop_out&#39;</span><span class="p">:</span> <span class="s1">&#39;0&#39;</span><span class="p">,</span>
        <span class="s1">&#39;weight_initial&#39;</span><span class="p">:</span> <span class="s1">&#39;glorot_uniform&#39;</span><span class="p">,</span>
        <span class="s1">&#39;weight_l1_penalty&#39;</span><span class="p">:</span><span class="s1">&#39;0.01&#39;</span><span class="p">,</span>
        <span class="s1">&#39;weight_decay&#39;</span><span class="p">:</span> <span class="s1">&#39;phylogenetic_tree&#39;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="s1">&#39;model_info&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;decay&#39;</span><span class="p">:</span> <span class="s1">&#39;0.001&#39;</span><span class="p">,</span>
        <span class="s1">&#39;loss&#39;</span><span class="p">:</span> <span class="s1">&#39;binary_crossentropy&#39;</span><span class="p">,</span>
        <span class="s1">&#39;lr&#39;</span><span class="p">:</span> <span class="s1">&#39;0.01&#39;</span><span class="p">,</span>
        <span class="s1">&#39;metrics&#39;</span><span class="p">:</span> <span class="s1">&#39;binary_accuracy, sensitivity, specificity, gmeasure, auc&#39;</span><span class="p">,</span>
        <span class="s1">&#39;network_class&#39;</span><span class="p">:</span> <span class="s1">&#39;DeepBiomeNetwork&#39;</span><span class="p">,</span>
        <span class="s1">&#39;normalizer&#39;</span><span class="p">:</span> <span class="s1">&#39;normalize_minmax&#39;</span><span class="p">,</span>
        <span class="s1">&#39;optimizer&#39;</span><span class="p">:</span> <span class="s1">&#39;adam&#39;</span><span class="p">,</span>
        <span class="s1">&#39;reader_class&#39;</span><span class="p">:</span> <span class="s1">&#39;MicroBiomeClassificationReader&#39;</span><span class="p">,</span>
        <span class="s1">&#39;taxa_selection_metrics&#39;</span><span class="p">:</span> <span class="s1">&#39;accuracy, sensitivity, specificity, gmeasure&#39;</span>
    <span class="p">},</span>
    <span class="s1">&#39;training_info&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;batch_size&#39;</span><span class="p">:</span> <span class="s1">&#39;200&#39;</span><span class="p">,</span>
        <span class="s1">&#39;epochs&#39;</span><span class="p">:</span> <span class="s1">&#39;10&#39;</span><span class="p">,</span>
        <span class="s1">&#39;callbacks&#39;</span><span class="p">:</span> <span class="s1">&#39;ModelCheckpoint&#39;</span><span class="p">,</span>
        <span class="s1">&#39;monitor&#39;</span><span class="p">:</span> <span class="s1">&#39;val_binary_accuracy&#39;</span><span class="p">,</span>
        <span class="s1">&#39;mode&#39;</span><span class="p">:</span> <span class="s1">&#39;max&#39;</span><span class="p">,</span>
        <span class="s1">&#39;min_delta&#39;</span><span class="p">:</span> <span class="s1">&#39;1e-4&#39;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="s1">&#39;validation_info&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;batch_size&#39;</span><span class="p">:</span> <span class="s1">&#39;None&#39;</span><span class="p">,</span> <span class="s1">&#39;validation_size&#39;</span><span class="p">:</span> <span class="s1">&#39;0.2&#39;</span>
    <span class="p">},</span>
    <span class="s1">&#39;test_info&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;batch_size&#39;</span><span class="p">:</span> <span class="s1">&#39;None&#39;</span>
    <span class="p">}</span>
<span class="p">}</span>
</pre></div>
</div>
<p>This is the example of the configuration file: <cite>network_info.cfg</cite></p>
<div class="highlight-cfg notranslate"><div class="highlight"><pre><span></span><span class="k">[model_info]</span>
<span class="na">network_class</span> <span class="o">=</span> <span class="s">DeepBiomeNetwork</span>
<span class="na">optimizer</span>   <span class="o">=</span> <span class="s">adam</span>
<span class="na">lr</span>          <span class="o">=</span> <span class="s">0.01</span>
<span class="na">decay</span>       <span class="o">=</span> <span class="s">0.0001</span>
<span class="na">loss</span>        <span class="o">=</span> <span class="s">binary_crossentropy</span>
<span class="na">metrics</span>     <span class="o">=</span> <span class="s">binary_accuracy, sensitivity, specificity, gmeasure, auc</span>
<span class="na">texa_selection_metrics</span> <span class="o">=</span> <span class="s">accuracy, sensitivity, specificity, gmeasure</span>
<span class="na">reader_class</span> <span class="o">=</span> <span class="s">MicroBiomeClassificationReader</span>
<span class="na">normalizer</span>  <span class="o">=</span> <span class="s">normalize_minmax</span>

<span class="k">[architecture_info]</span>
<span class="na">weight_initial</span> <span class="o">=</span> <span class="s">glorot_uniform</span>
<span class="na">weight_decay</span> <span class="o">=</span> <span class="s">phylogenetic_tree</span>
<span class="na">batch_normalization</span> <span class="o">=</span> <span class="s">False</span>
<span class="na">drop_out</span> <span class="o">=</span> <span class="s">0</span>

<span class="k">[training_info]</span>
<span class="na">epochs</span>          <span class="o">=</span> <span class="s">1000</span>
<span class="na">batch_size</span>      <span class="o">=</span> <span class="s">200</span>
<span class="na">callbacks</span>       <span class="o">=</span> <span class="s">ModelCheckpoint</span>
<span class="na">monitor</span>         <span class="o">=</span> <span class="s">val_binary_accuracy</span>
<span class="na">mode</span>            <span class="o">=</span> <span class="s">max</span>
<span class="na">min_delta</span>       <span class="o">=</span> <span class="s">1e-4</span>

<span class="k">[validation_info]</span>
<span class="na">validation_size</span> <span class="o">=</span> <span class="s">0.2</span>
<span class="na">batch_size</span> <span class="o">=</span> <span class="s">None</span>

<span class="k">[test_info]</span>
<span class="na">batch_size</span> <span class="o">=</span> <span class="s">None</span>
</pre></div>
</div>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>See <a class="reference external" href="https://github.com/Young-won/deepbiome/tree/master/examples">Example</a> for reference about the configuration file example for various problems.</p>
</div>
</div>
</div>
<div class="section" id="preparing-the-configuration-about-the-path-information-path-info">
<h3>Preparing the configuration about the path information (<cite>path_info</cite>)<a class="headerlink" href="#preparing-the-configuration-about-the-path-information-path-info" title="Permalink to this headline">¶</a></h3>
<p>To give the information about the path to dataset, paths for saving the trained weights and the evaluation results, we provide a dictionary of configurations to the <cite>path_info</cite> feild. Alternatively we can also use the configufation file (.cfg).</p>
<p>Your configuration for the paths should include the information about:</p>
<dl class="field-list simple">
<dt class="field-odd">data_info</dt>
<dd class="field-odd"><p>about the path information of the dataset</p>
</dd>
<dt class="field-even">model_info</dt>
<dd class="field-even"><p>about the path information for saving the trained weights and the evaluation results</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>All paths are the relative path based on the directory where code will run.</p>
</div>
<div class="section" id="path-info-data-info">
<h4>path_info[‘data_info’]<a class="headerlink" href="#path-info-data-info" title="Permalink to this headline">¶</a></h4>
<p>To provide the dictionary as input, we can use the option below:</p>
<dl class="field-list simple">
<dt class="field-odd">tree_info_path</dt>
<dd class="field-odd"><p>tree information file (.csv)</p>
</dd>
<dt class="field-even">count_list_path</dt>
<dd class="field-even"><p>lists of the name of input files (.csv)</p>
</dd>
<dt class="field-odd">count_path</dt>
<dd class="field-odd"><p>directory path of the input files</p>
</dd>
<dt class="field-even">y_path</dt>
<dd class="field-even"><p>y path (.csv)  (not required for prediction)</p>
</dd>
<dt class="field-odd">idx_path</dt>
<dd class="field-odd"><p>index path for repetation (.csv)</p>
</dd>
<dt class="field-even">data_path</dt>
<dd class="field-even"><p>directory path of the index and y file</p>
</dd>
</dl>
<p>To provide one configuration file, we can use the options below:</p>
<dl class="field-list simple">
<dt class="field-odd">tree_info_path</dt>
<dd class="field-odd"><p>tree information file (.csv)</p>
</dd>
<dt class="field-even">x_path</dt>
<dd class="field-even"><p>input path (.csv)</p>
</dd>
<dt class="field-odd">y_path</dt>
<dd class="field-odd"><p>y path (.csv)  (not required for prediction)</p>
</dd>
<dt class="field-even">data_path</dt>
<dd class="field-even"><p>directory path of the index, x and y file</p>
</dd>
</dl>
</div>
<div class="section" id="path-info-model-info">
<h4>path_info[‘model_info’]<a class="headerlink" href="#path-info-model-info" title="Permalink to this headline">¶</a></h4>
<dl class="field-list simple">
<dt class="field-odd">weight</dt>
<dd class="field-odd"><p>weight file name (.h5)</p>
</dd>
<dt class="field-even">evaluation</dt>
<dd class="field-even"><p>evaluation file name (.npy)  (not required for prediction)</p>
</dd>
<dt class="field-odd">model_dir</dt>
<dd class="field-odd"><p>base directory path for the model (weight, evaluation)</p>
</dd>
<dt class="field-even">history</dt>
<dd class="field-even"><p>history file name for the history value of each evaluation metric from the training (.json). If not setted, <cite>deepbiome</cite> will not save the history of the network training.</p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>If you want to use sub-directories in the path (for example, “weight”=”weight/weight.h5”, “history”=”history/hist.h5”, “model_dir”=”./”), you should have to make the sub-directories “./weight” and “./history” before running the code.</p>
</div>
</div>
<div class="section" id="example-for-the-path-info-for-the-list-of-inputs">
<h4>Example for the <cite>path_info</cite> for the list of inputs<a class="headerlink" href="#example-for-the-path-info-for-the-list-of-inputs" title="Permalink to this headline">¶</a></h4>
<p>This is the example of the configuration dictionary: <cite>path_info</cite> dictionary</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">path_info</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;data_info&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;count_list_path&#39;</span><span class="p">:</span> <span class="s1">&#39;data/simulation/gcount_list.csv&#39;</span><span class="p">,</span>
        <span class="s1">&#39;count_path&#39;</span><span class="p">:</span> <span class="s1">&#39;data/simulation/count&#39;</span><span class="p">,</span>
        <span class="s1">&#39;data_path&#39;</span><span class="p">:</span> <span class="s1">&#39;data/simulation/s2/&#39;</span><span class="p">,</span>
        <span class="s1">&#39;idx_path&#39;</span><span class="p">:</span> <span class="s1">&#39;data/simulation/s2/idx.csv&#39;</span><span class="p">,</span>
        <span class="s1">&#39;tree_info_path&#39;</span><span class="p">:</span> <span class="s1">&#39;data/genus48/genus48_dic.csv&#39;</span><span class="p">,</span>
        <span class="s1">&#39;x_path&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
        <span class="s1">&#39;y_path&#39;</span><span class="p">:</span> <span class="s1">&#39;y.csv&#39;</span>
    <span class="p">},</span>
    <span class="s1">&#39;model_info&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;model_dir&#39;</span><span class="p">:</span> <span class="s1">&#39;./simulation_s2/simulation_s2_deepbiome/&#39;</span><span class="p">,</span>
        <span class="s1">&#39;weight&#39;</span><span class="p">:</span> <span class="s1">&#39;weight/weight.h5&#39;</span><span class="p">,</span>
        <span class="s1">&#39;history&#39;</span><span class="p">:</span> <span class="s1">&#39;hist.json&#39;</span><span class="p">,</span>
        <span class="s1">&#39;evaluation&#39;</span><span class="p">:</span> <span class="s1">&#39;eval.npy&#39;</span>
    <span class="p">}</span>
<span class="p">}</span>
</pre></div>
</div>
<p>This is the example of the configuration file: <cite>path_info.cfg</cite></p>
<div class="highlight-cfg notranslate"><div class="highlight"><pre><span></span><span class="k">[data_info]</span>
<span class="na">data_path</span> <span class="o">=</span> <span class="s">data/simulation/s2/</span>
<span class="na">tree_info_path</span> <span class="o">=</span> <span class="s">data/genus48/genus48_dic.csv</span>
<span class="na">idx_path</span> <span class="o">=</span> <span class="s">data/simulation/s2/idx.csv</span>
<span class="na">count_list_path</span> <span class="o">=</span> <span class="s">data/simulation/gcount_list.csv</span>
<span class="na">count_path</span> <span class="o">=</span> <span class="s">data/simulation/count</span>
<span class="na">y_path</span> <span class="o">=</span> <span class="s">y.csv</span>

<span class="k">[model_info]</span>
<span class="na">model_dir</span> <span class="o">=</span> <span class="s">./simulation_s2/simulation_s2_deepbiome/</span>
<span class="na">weight</span> <span class="o">=</span> <span class="s">weight/weight.h5</span>
<span class="na">history</span> <span class="o">=</span> <span class="s">historys/hist.json</span>
<span class="na">evaluation</span> <span class="o">=</span> <span class="s">eval.npy</span>
</pre></div>
</div>
</div>
<div class="section" id="example-for-the-path-info-for-the-one-input-file">
<h4>Example for the <cite>path_info</cite> for the one input file<a class="headerlink" href="#example-for-the-path-info-for-the-one-input-file" title="Permalink to this headline">¶</a></h4>
<p>This is the example of the configuration dictionary: <cite>path_info</cite> dictionary</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">path_info</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;data_info&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;data_path&#39;</span><span class="p">:</span> <span class="s1">&#39;../../data/pulmonary/&#39;</span><span class="p">,</span>
        <span class="s1">&#39;tree_info_path&#39;</span><span class="p">:</span> <span class="s1">&#39;../../data/genus48/genus48_dic.csv&#39;</span><span class="p">,</span>
        <span class="s1">&#39;x_path&#39;</span><span class="p">:</span> <span class="s1">&#39;X.csv&#39;</span><span class="p">,</span>
        <span class="s1">&#39;y_path&#39;</span><span class="p">:</span> <span class="s1">&#39;y.csv&#39;</span>
    <span class="p">},</span>
    <span class="s1">&#39;model_info&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;model_dir&#39;</span><span class="p">:</span> <span class="s1">&#39;./&#39;</span><span class="p">,</span>
        <span class="s1">&#39;weight&#39;</span><span class="p">:</span> <span class="s1">&#39;weight/weight.h5&#39;</span><span class="p">,</span>
        <span class="s1">&#39;history&#39;</span><span class="p">:</span><span class="s1">&#39;history/hist.json&#39;</span><span class="p">,</span>
        <span class="s1">&#39;evaluation&#39;</span><span class="p">:</span> <span class="s1">&#39;eval.npy&#39;</span><span class="p">,</span>
    <span class="p">}</span>
<span class="p">}</span>
</pre></div>
</div>
<p>This is the example of the configuration file: <cite>path_info.cfg</cite></p>
<div class="highlight-cfg notranslate"><div class="highlight"><pre><span></span><span class="k">[data_info]</span>
<span class="na">data_path</span> <span class="o">=</span> <span class="s">../../data/pulmonary/</span>
<span class="na">tree_info_path</span> <span class="o">=</span> <span class="s">../../data/genus48/genus48_dic.csv</span>
<span class="na">x_path</span> <span class="o">=</span> <span class="s">X.csv</span>
<span class="na">y_path</span> <span class="o">=</span> <span class="s">y.csv</span>

<span class="k">[model_info]</span>
<span class="na">model_dir</span> <span class="o">=</span> <span class="s">./</span>
<span class="na">weight</span> <span class="o">=</span> <span class="s">weight/weight.h5</span>
<span class="na">history</span> <span class="o">=</span> <span class="s">history/hist.json</span>
<span class="na">evaluation</span> <span class="o">=</span> <span class="s">eval.npy</span>
</pre></div>
</div>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>See <a class="reference external" href="https://github.com/Young-won/deepbiome/tree/master/examples">Example</a> for reference about the configuration file example for various problems.</p>
</div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="usage.html" class="btn btn-neutral float-right" title="Usage" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Youngwon Choi

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>