

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Usage &mdash; deepbiome 0.1.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script src="_static/clipboard.min.js"></script>
        <script src="_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"processClass": "math|output_area", "inlineMath": [["$", "$"], ["\\(", "\\)"]], "ignoreClass": "document", "processEscapes": true}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/copybutton.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Example : k fold cross-validation with an input file" href="example_with_one_input_file.html" />
    <link rel="prev" title="Prerequisites" href="prerequisites.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> deepbiome
          

          
          </a>

          
            
            
              <div class="version">
                0.1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="DeepBiome_usage_classification.html">Getting start with the classification problem.</a></li>
<li class="toctree-l1"><a class="reference internal" href="DeepBiome_usage_regression.html">Getting start with the regression problem.</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="prerequisites.html">Prerequisites</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#preparing-the-dataset">Preparing the dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="#preparing-the-configufation">Preparing the configufation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#using-the-configuration-dictionary">Using the configuration dictionary</a></li>
<li class="toctree-l3"><a class="reference internal" href="#using-the-configuration-file-cfg">Using the configuration file (.cfg)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#logging">Logging</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#training">Training</a></li>
<li class="toctree-l2"><a class="reference internal" href="#testing">Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="#prediction">Prediction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#cheatsheet-for-running-the-project-on-console">Cheatsheet for running the project on console</a></li>
<li class="toctree-l2"><a class="reference internal" href="#summary">Summary</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="example_with_one_input_file.html">Example : k fold cross-validation with an input file</a></li>
<li class="toctree-l1"><a class="reference internal" href="example_with_the_list_of_inputs.html">Example : k times repetition with the list of k input files</a></li>
<li class="toctree-l1"><a class="reference internal" href="release-history.html">Release History</a></li>
<li class="toctree-l1"><a class="reference internal" href="min_versions.html">Minimum Version of Python and NumPy</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">deepbiome</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Usage</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/usage.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h1>
<div class="section" id="preparing-the-dataset">
<h2>Preparing the dataset<a class="headerlink" href="#preparing-the-dataset" title="Permalink to this headline">¶</a></h2>
<p>We assume that you have the dataset</p>
<p>See <a class="reference internal" href="prerequisites.html#data-preprocessing"><span class="std std-ref">Data preprocessing</span></a>.</p>
</div>
<div class="section" id="preparing-the-configufation">
<h2>Preparing the configufation<a class="headerlink" href="#preparing-the-configufation" title="Permalink to this headline">¶</a></h2>
<p>To provide configurations for the training process and path information, we can use either dictionaries as input to functions or configuration files.</p>
<div class="section" id="using-the-configuration-dictionary">
<h3>Using the configuration dictionary<a class="headerlink" href="#using-the-configuration-dictionary" title="Permalink to this headline">¶</a></h3>
<p>You can use the network configuration dictionary <cite>network_info</cite> and the path configuration dictionary <cite>path_info</cite> for configuration.</p>
<p>See <a class="reference internal" href="prerequisites.html#configuration"><span class="std std-ref">Configuration</span></a>.</p>
</div>
<div class="section" id="using-the-configuration-file-cfg">
<h3>Using the configuration file (.cfg)<a class="headerlink" href="#using-the-configuration-file-cfg" title="Permalink to this headline">¶</a></h3>
<p>You can use the configuration format files <cite>.cfg</cite> for configuration (for example, <cite>path_info.cfg</cite> and <cite>network_info.cfg</cite>).</p>
<p>See <a class="reference internal" href="prerequisites.html#configuration"><span class="std std-ref">Configuration</span></a>.</p>
<p>If then, you can get the dictionaries from the config files by:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># for python 3.x</span>
<span class="kn">from</span> <span class="nn">deepbiome</span> <span class="kn">import</span> <span class="n">configuration</span>

<span class="n">config_data</span> <span class="o">=</span> <span class="n">configuration</span><span class="o">.</span><span class="n">Configurator</span><span class="p">(</span><span class="s1">&#39;./path_info.cfg&#39;</span><span class="p">,</span> <span class="n">log</span><span class="p">)</span>
<span class="n">config_data</span><span class="o">.</span><span class="n">set_config_map</span><span class="p">(</span><span class="n">config_data</span><span class="o">.</span><span class="n">get_section_map</span><span class="p">())</span>
<span class="n">config_data</span><span class="o">.</span><span class="n">print_config_map</span><span class="p">()</span>
<span class="n">path_info</span> <span class="o">=</span> <span class="n">config_data</span><span class="o">.</span><span class="n">get_config_map</span><span class="p">()</span>

<span class="n">config_network</span> <span class="o">=</span> <span class="n">configuration</span><span class="o">.</span><span class="n">Configurator</span><span class="p">(</span><span class="s1">&#39;./network_info.cfg&#39;</span><span class="p">,</span> <span class="n">log</span><span class="p">)</span>
<span class="n">config_network</span><span class="o">.</span><span class="n">set_config_map</span><span class="p">(</span><span class="n">config_network</span><span class="o">.</span><span class="n">get_section_map</span><span class="p">())</span>
<span class="n">config_network</span><span class="o">.</span><span class="n">print_config_map</span><span class="p">()</span>
<span class="n">network_info</span> <span class="o">=</span> <span class="n">config_network</span><span class="o">.</span><span class="n">get_config_map</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="logging">
<h3>Logging<a class="headerlink" href="#logging" title="Permalink to this headline">¶</a></h3>
<p>For logging, we can use the logging instance. For example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># for python 3.x</span>
<span class="kn">import</span> <span class="nn">logging</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="nb">format</span> <span class="o">=</span> <span class="s1">&#39;[</span><span class="si">%(name)-8s</span><span class="s1">|</span><span class="si">%(levelname)s</span><span class="s1">|</span><span class="si">%(filename)s</span><span class="s1">:</span><span class="si">%(lineno)s</span><span class="s1">] </span><span class="si">%(message)s</span><span class="s1">&#39;</span><span class="p">,</span>
                <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">DEBUG</span><span class="p">)</span>

<span class="n">log</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">()</span>
</pre></div>
</div>
<p>We can use the logging instance <cite>log</cite> as an input of the main training fuction below.</p>
<p>For more information about <cite>logging</cite> module, please check the documentation <a class="reference external" href="https://docs.python.org/3/library/logging.html">logging</a>.</p>
</div>
</div>
<div class="section" id="training">
<h2>Training<a class="headerlink" href="#training" title="Permalink to this headline">¶</a></h2>
<p>To use DeepBiome:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># for python 3.x</span>
<span class="kn">from</span> <span class="nn">deepbiome</span> <span class="kn">import</span> <span class="n">deepbiome</span>

<span class="n">test_evaluation</span><span class="p">,</span> <span class="n">train_evaluation</span><span class="p">,</span> <span class="n">network</span> <span class="o">=</span> <span class="n">deepbiome</span><span class="o">.</span><span class="n">deepbiome_train</span><span class="p">(</span><span class="n">log</span><span class="p">,</span> <span class="n">network_info</span><span class="p">,</span> <span class="n">path_info</span><span class="p">)</span>
</pre></div>
</div>
<p>If you want to train the network with specific number <cite>k</cite> of cross-validation, you can set the <cite>number_of_fold</cite>. For example, if you want to run the 5-fold cross-validation:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># for python 3.x</span>
<span class="kn">from</span> <span class="nn">deepbiome</span> <span class="kn">import</span> <span class="n">deepbiome</span>

<span class="n">test_evaluation</span><span class="p">,</span> <span class="n">train_evaluation</span><span class="p">,</span> <span class="n">network</span> <span class="o">=</span> <span class="n">deepbiome</span><span class="o">.</span><span class="n">deepbiome_train</span><span class="p">(</span><span class="n">log</span><span class="p">,</span> <span class="n">network_info</span><span class="p">,</span> <span class="n">path_info</span><span class="p">,</span> <span class="n">number_of_fold</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
<p>If you use one input file, it will run the 5-fold cross validation using all data in that file. If you use the list of the input files, it will run the training by the first <cite>k</cite> files.</p>
<p>By defaults, <cite>number_of_fold=None</cite>. Then, the code will run for the leave-one-out-cross-validation (LOOCV).</p>
<p>If you use one input file, it will run the LOOCV using all data in that file. If you use the list of the input files, it will repeat the training for every file.</p>
</div>
<div class="section" id="testing">
<h2>Testing<a class="headerlink" href="#testing" title="Permalink to this headline">¶</a></h2>
<p>If you want to test the pre-trained model, you can use the <cite>deepbiome.deepbiome_test</cite> function.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># for python 3.x</span>
<span class="kn">from</span> <span class="nn">deepbiome</span> <span class="kn">import</span> <span class="n">deepbiome</span>

<span class="n">evaluation</span> <span class="o">=</span> <span class="n">deepbiome</span><span class="o">.</span><span class="n">deepbiome_test</span><span class="p">(</span><span class="n">log</span><span class="p">,</span> <span class="n">network_info</span><span class="p">,</span> <span class="n">path_info</span><span class="p">,</span> <span class="n">number_of_fold</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
<p>If you use the index file, this function provides the evaluation using test index (index set not included in the index file) for each fold. If not, this function provides the evaluation using the whole sample.
If <cite>number_of_fold</cite> is setted as <cite>k</cite>, the function will test the model only with first <cite>k</cite> folds.</p>
<p>This function provides the evaluation result as a numpy array with a shape of (number of folds, number of evaluation measures).</p>
</div>
<div class="section" id="prediction">
<h2>Prediction<a class="headerlink" href="#prediction" title="Permalink to this headline">¶</a></h2>
<p>If you want to predict the output using the pre-trained model, you can use the <cite>deepbiome.deepbiome_prediction</cite> function.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># for python 3.x</span>
<span class="kn">from</span> <span class="nn">deepbiome</span> <span class="kn">import</span> <span class="n">deepbiome</span>

<span class="n">evaluation</span> <span class="o">=</span> <span class="n">deepbiome</span><span class="o">.</span><span class="n">deepbiome_prediction</span><span class="p">(</span><span class="n">log</span><span class="p">,</span> <span class="n">network_info</span><span class="p">,</span> <span class="n">path_info</span><span class="p">,</span> <span class="n">num_classes</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">number_of_fold</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">change_weight_for_each_fold</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>If <cite>number_of_fold</cite> is set as <cite>k</cite>, the function will predict the output of the first <cite>k</cite> folds’ samples.</p>
<p>If <cite>change_weight_for_each_fold</cite> is set as <cite>False</cite>, the function will predict the output of every repeatition by same weight from the given path.
If <cite>change_weight_for_each_fold</cite> is set as <cite>True</cite>, the function will predict the output of by each fold weight.</p>
<p>If ‘get_y=True’, the function will provide a list of tuples (prediction, true output) as a numpy array output with the shape of <cite>(n_samples, 2, n_classes)</cite>. If ‘get_y=False’, the function will provide a numpy array of predictions only. The numpy array output will have the shape of <cite>(n_samples, n_classes)</cite>.</p>
</div>
<div class="section" id="cheatsheet-for-running-the-project-on-console">
<h2>Cheatsheet for running the project on console<a class="headerlink" href="#cheatsheet-for-running-the-project-on-console" title="Permalink to this headline">¶</a></h2>
<ol class="arabic">
<li><dl class="simple">
<dt>Preprocessing the data (convert raw data to the format readable for python): (See <a class="reference internal" href="prerequisites.html#data-preprocessing"><span class="std std-ref">Data preprocessing</span></a>.)</dt><dd><p>Example: TODO (Julia)</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Set configuration file about training hyperparameter and path information:</dt><dd><ol class="arabic simple">
<li><dl class="simple">
<dt>Set the training hyper-parameter (<cite>network_info.cfg</cite>): (See <a class="reference internal" href="prerequisites.html#configuration"><span class="std std-ref">Configuration</span></a>.)</dt><dd><p>Example: <a class="reference external" href="https://github.com/Young-won/deepbiome/tree/master/examples/simulation_s0/simulation_s0_deepbiome//config/network_info.cfg">https://github.com/Young-won/deepbiome/tree/master/examples/simulation_s0/simulation_s0_deepbiome//config/network_info.cfg</a></p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Set the path information (<cite>path_info.cfg</cite>): (See <a class="reference internal" href="prerequisites.html#configuration"><span class="std std-ref">Configuration</span></a>.)</dt><dd><p>Example: <a class="reference external" href="https://github.com/Young-won/deepbiome/tree/master/examples/simulation_s0/simulation_s0_deepbiome/config/path_info.cfg">https://github.com/Young-won/deepbiome/tree/master/examples/simulation_s0/simulation_s0_deepbiome/config/path_info.cfg</a></p>
</dd>
</dl>
</li>
</ol>
</dd>
</dl>
</li>
<li><dl>
<dt>Write the python script for running the function <cite>deepbiome.deepbiome_train</cite>. For example:</dt><dd><p>Example: <a class="reference external" href="https://github.com/Young-won/deepbiome/tree/master/examples/main.py">https://github.com/Young-won/deepbiome/tree/master/examples/main.py</a></p>
<p>Example of the python script:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sys</span>

<span class="kn">from</span> <span class="nn">deepbiome</span> <span class="kn">import</span> <span class="n">configuration</span>
<span class="kn">from</span> <span class="nn">deepbiome</span> <span class="kn">import</span> <span class="n">logging_daily</span>
<span class="kn">from</span> <span class="nn">deepbiome</span> <span class="kn">import</span> <span class="n">deepbiome</span>
<span class="kn">from</span> <span class="nn">deepbiome.utils</span> <span class="kn">import</span> <span class="n">argv_parse</span>

<span class="c1"># Argument ##########################################################</span>
<span class="n">argdict</span> <span class="o">=</span> <span class="n">argv_parse</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">)</span>
<span class="k">try</span><span class="p">:</span> <span class="n">gpu_memory_fraction</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">argdict</span><span class="p">[</span><span class="s1">&#39;gpu_memory_fraction&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
<span class="k">except</span><span class="p">:</span> <span class="n">gpu_memory_fraction</span> <span class="o">=</span> <span class="kc">None</span>
<span class="k">try</span><span class="p">:</span> <span class="n">max_queue_size</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">argdict</span><span class="p">[</span><span class="s1">&#39;max_queue_size&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
<span class="k">except</span><span class="p">:</span> <span class="n">max_queue_size</span><span class="o">=</span><span class="mi">10</span>
<span class="k">try</span><span class="p">:</span> <span class="n">workers</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">argdict</span><span class="p">[</span><span class="s1">&#39;workers&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
<span class="k">except</span><span class="p">:</span> <span class="n">workers</span><span class="o">=</span><span class="mi">1</span>
<span class="k">try</span><span class="p">:</span> <span class="n">use_multiprocessing</span><span class="o">=</span><span class="n">argdict</span><span class="p">[</span><span class="s1">&#39;use_multiprocessing&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;True&#39;</span>
<span class="k">except</span><span class="p">:</span> <span class="n">use_multiprocessing</span><span class="o">=</span><span class="kc">False</span>

<span class="c1"># Logger ###########################################################</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging_daily</span><span class="o">.</span><span class="n">logging_daily</span><span class="p">(</span><span class="n">argdict</span><span class="p">[</span><span class="s1">&#39;log_info&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
<span class="n">logger</span><span class="o">.</span><span class="n">reset_logging</span><span class="p">()</span>
<span class="n">log</span> <span class="o">=</span> <span class="n">logger</span><span class="o">.</span><span class="n">get_logging</span><span class="p">()</span>
<span class="n">log</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging_daily</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>

<span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Argument input&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">argname</span><span class="p">,</span> <span class="n">arg</span> <span class="ow">in</span> <span class="n">argdict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;    </span><span class="si">{}</span><span class="s1">:</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">argname</span><span class="p">,</span><span class="n">arg</span><span class="p">))</span>

<span class="c1"># Configuration ####################################################</span>
<span class="n">config_data</span> <span class="o">=</span> <span class="n">configuration</span><span class="o">.</span><span class="n">Configurator</span><span class="p">(</span><span class="n">argdict</span><span class="p">[</span><span class="s1">&#39;path_info&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">log</span><span class="p">)</span>
<span class="n">config_data</span><span class="o">.</span><span class="n">set_config_map</span><span class="p">(</span><span class="n">config_data</span><span class="o">.</span><span class="n">get_section_map</span><span class="p">())</span>
<span class="n">config_data</span><span class="o">.</span><span class="n">print_config_map</span><span class="p">()</span>

<span class="n">config_network</span> <span class="o">=</span> <span class="n">configuration</span><span class="o">.</span><span class="n">Configurator</span><span class="p">(</span><span class="n">argdict</span><span class="p">[</span><span class="s1">&#39;network_info&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">log</span><span class="p">)</span>
<span class="n">config_network</span><span class="o">.</span><span class="n">set_config_map</span><span class="p">(</span><span class="n">config_network</span><span class="o">.</span><span class="n">get_section_map</span><span class="p">())</span>
<span class="n">config_network</span><span class="o">.</span><span class="n">print_config_map</span><span class="p">()</span>

<span class="n">path_info</span> <span class="o">=</span> <span class="n">config_data</span><span class="o">.</span><span class="n">get_config_map</span><span class="p">()</span>
<span class="n">network_info</span> <span class="o">=</span> <span class="n">config_network</span><span class="o">.</span><span class="n">get_config_map</span><span class="p">()</span>
<span class="n">test_evaluation</span><span class="p">,</span> <span class="n">train_evaluation</span><span class="p">,</span> <span class="n">network</span> <span class="o">=</span> <span class="n">deepbiome</span><span class="o">.</span><span class="n">deepbiome_train</span><span class="p">(</span><span class="n">log</span><span class="p">,</span> <span class="n">network_info</span><span class="p">,</span> <span class="n">path_info</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</li>
</ol>
<ol class="arabic" start="5">
<li><p>Check the available GPU (if you have no GPU, it will run on CPU):</p>
<blockquote>
<div><div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nvidia-smi</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><dl>
<dt>Select the number of GPUs and CPU cores from the bash file:</dt><dd><p>Example: <a class="reference external" href="https://github.com/Young-won/deepbiome/tree/master/examples/simulation_s0/simulation_s0_deepbiome/run.sh">https://github.com/Young-won/deepbiome/tree/master/examples/simulation_s0/simulation_s0_deepbiome/run.sh</a></p>
<p>Example of the bash file (.sh):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">export</span> <span class="nv">CUDA_VISIBLE_DEVICES</span><span class="o">=</span><span class="m">2</span>
<span class="nb">echo</span> <span class="nv">$CUDA_VISIBLE_DEVICES</span>

<span class="nv">model</span><span class="o">=</span><span class="si">${</span><span class="nv">PWD</span><span class="p">##*/</span><span class="si">}</span>
<span class="nb">echo</span> <span class="nv">$model</span>

python3 ../../main.py --log_info<span class="o">=</span>config/log_info.yaml --path_info<span class="o">=</span>config/path_info.cfg --network_info<span class="o">=</span>config/network_info.cfg  --max_queue_size<span class="o">=</span><span class="m">50</span> --workers<span class="o">=</span><span class="m">10</span> --use_multiprocessing<span class="o">=</span>False
</pre></div>
</div>
</dd>
</dl>
</li>
<li><p>Run the bash file!</p>
<blockquote>
<div><div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">./run.sh</span>
</pre></div>
</div>
</div></blockquote>
</li>
</ol>
</div>
<div class="section" id="summary">
<h2>Summary<a class="headerlink" href="#summary" title="Permalink to this headline">¶</a></h2>
<p>To use deepbiome in a project:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>from deepbiome import deepbiome
</pre></div>
</div>
<dl class="function">
<dt id="deepbiome.deepbiome.deepbiome_train">
<code class="sig-prename descclassname">deepbiome.deepbiome.</code><code class="sig-name descname">deepbiome_train</code><span class="sig-paren">(</span><em class="sig-param">log, network_info, path_info, number_of_fold=None, tree_level_list=['Genus', 'Family', 'Order', 'Class', 'Phylum'], max_queue_size=10, workers=1, use_multiprocessing=False, verbose=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/deepbiome/deepbiome.html#deepbiome_train"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#deepbiome.deepbiome.deepbiome_train" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for training the deep neural network with phylogenetic tree weight regularizer.</p>
<p>It uses microbiome abundance data as input and uses the phylogenetic taxonomy to guide the decision of the optimal number of layers and neurons in the deep learning architecture.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>log (logging instance) :</strong></dt><dd><p>python logging instance for logging</p>
</dd>
<dt><strong>network_info (dictionary) :</strong></dt><dd><p>python dictionary with network_information</p>
</dd>
<dt><strong>path_info (dictionary):</strong></dt><dd><p>python dictionary with path_information</p>
</dd>
<dt><strong>number_of_fold (int):</strong></dt><dd><p>default=None</p>
</dd>
<dt><strong>tree_level_list (list):</strong></dt><dd><p>name of each level of the given reference tree weights
default=[‘Genus’, ‘Family’, ‘Order’, ‘Class’, ‘Phylum’]</p>
</dd>
<dt><strong>max_queue_size (int):</strong></dt><dd><p>default=10</p>
</dd>
<dt><strong>workers (int):</strong></dt><dd><p>default=1</p>
</dd>
<dt><strong>use_multiprocessing (boolean):</strong></dt><dd><p>default=False</p>
</dd>
<dt><strong>verbose (boolean):</strong></dt><dd><p>show the log if True
default=True</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>test_evaluation (numpy array):</dt><dd><p>numpy array of the evaluation using testset from all fold</p>
</dd>
<dt>train_evaluation (numpy array):</dt><dd><p>numpy array of the evaluation using training from all fold</p>
</dd>
<dt>network (deepbiome network instance):</dt><dd><p>deepbiome class instance</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Training the deep neural network with phylogenetic tree weight regularizer.</p>
<p>test_evaluation, train_evaluation, network = deepbiome_train(log, network_info, path_info)</p>
</dd></dl>

<dl class="function">
<dt id="deepbiome.deepbiome.deepbiome_test">
<code class="sig-prename descclassname">deepbiome.deepbiome.</code><code class="sig-name descname">deepbiome_test</code><span class="sig-paren">(</span><em class="sig-param">log, network_info, path_info, number_of_fold=None, tree_level_list=['Genus', 'Family', 'Order', 'Class', 'Phylum'], max_queue_size=10, workers=1, use_multiprocessing=False, verbose=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/deepbiome/deepbiome.html#deepbiome_test"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#deepbiome.deepbiome.deepbiome_test" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for testing the pretrained deep neural network with phylogenetic tree weight regularizer.</p>
<p>If you use the index file, this function provide the evaluation using test index (index set not included in the index file) for each fold. If not, this function provide the evaluation using the whole samples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>log (logging instance) :</strong></dt><dd><p>python logging instance for logging</p>
</dd>
<dt><strong>network_info (dictionary) :</strong></dt><dd><p>python dictionary with network_information</p>
</dd>
<dt><strong>path_info (dictionary):</strong></dt><dd><p>python dictionary with path_information</p>
</dd>
<dt><strong>number_of_fold (int):</strong></dt><dd><p>If <cite>number_of_fold</cite> is set as <cite>k</cite>, the function will test the model only with first <cite>k</cite> folds.
default=None</p>
</dd>
<dt><strong>tree_level_list (list):</strong></dt><dd><p>name of each level of the given reference tree weights
default=[‘Genus’, ‘Family’, ‘Order’, ‘Class’, ‘Phylum’]</p>
</dd>
<dt><strong>max_queue_size (int):</strong></dt><dd><p>default=10</p>
</dd>
<dt><strong>workers (int):</strong></dt><dd><p>default=1</p>
</dd>
<dt><strong>use_multiprocessing (boolean):</strong></dt><dd><p>default=False</p>
</dd>
<dt><strong>verbose (boolean):</strong></dt><dd><p>show the log if True
default=True</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>evaluation (numpy array):</dt><dd><p>evaluation result using testset as a numpy array with a shape of (number of fold, number of evaluation measures)</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Test the pre-trained deep neural network with phylogenetic tree weight regularizer.</p>
<p>evaluation = deepbiome_test(log, network_info, path_info)</p>
</dd></dl>

<dl class="function">
<dt id="deepbiome.deepbiome.deepbiome_prediction">
<code class="sig-prename descclassname">deepbiome.deepbiome.</code><code class="sig-name descname">deepbiome_prediction</code><span class="sig-paren">(</span><em class="sig-param">log, network_info, path_info, num_classes, number_of_fold=None, change_weight_for_each_fold=False, get_y=False, tree_level_list=['Genus', 'Family', 'Order', 'Class', 'Phylum'], max_queue_size=10, workers=1, use_multiprocessing=False, verbose=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/deepbiome/deepbiome.html#deepbiome_prediction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#deepbiome.deepbiome.deepbiome_prediction" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for prediction by the pretrained deep neural network with phylogenetic tree weight regularizer.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>log (logging instance) :</strong></dt><dd><p>python logging instance for logging</p>
</dd>
<dt><strong>network_info (dictionary) :</strong></dt><dd><p>python dictionary with network_information</p>
</dd>
<dt><strong>path_info (dictionary):</strong></dt><dd><p>python dictionary with path_information</p>
</dd>
<dt><strong>num_classes (int):</strong></dt><dd><p>number of classes for the network. 0 for regression, 1 for binary classificatin.</p>
</dd>
<dt><strong>number_of_fold (int):</strong></dt><dd><ol class="arabic simple">
<li><p>For the list of input files for repeatitions, the function will predict the output of the first <cite>number_of_fold</cite> repetitions. If <cite>number_of_fold</cite> is None, then the function will predict the output of the whole repetitions.</p></li>
<li><p>For the one input file for cross-validation, the function will predict the output of the <cite>k</cite>-fold cross validatoin. If <cite>number_of_fold</cite> is None, then the function will predict the output of the LOOCV.</p></li>
</ol>
<p>default=None</p>
</dd>
<dt><strong>change_weight_for_each_fold (boolean):</strong></dt><dd><p>If <cite>True</cite>, weight will be changed for each fold (repetition). For example, if the given weight’s name is <cite>weight.h5</cite> then <cite>weight_0.h5</cite> will loaded for the first fold (repetition). If <cite>False</cite>, weight path in the path_info will used for whole prediction. For example, if the given weight’s name is <cite>weight_0.h5</cite> then <cite>weight_0.h5</cite> will used for whole fold (repetition).
default=False</p>
</dd>
<dt><strong>get_y (boolean):</strong></dt><dd><p>If ‘True’, the function will provide a list of tuples (prediction, true output) as a output.
degault=False</p>
</dd>
<dt><strong>tree_level_list (list):</strong></dt><dd><p>name of each level of the given reference tree weights
default=[‘Genus’, ‘Family’, ‘Order’, ‘Class’, ‘Phylum’]</p>
</dd>
<dt><strong>max_queue_size (int):</strong></dt><dd><p>default=10</p>
</dd>
<dt><strong>workers (int):</strong></dt><dd><p>default=1</p>
</dd>
<dt><strong>use_multiprocessing (boolean):</strong></dt><dd><p>default=False</p>
</dd>
<dt><strong>verbose (boolean):</strong></dt><dd><p>show the log if True
default=True</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>prediction (numpy array):</dt><dd><p>prediction using whole dataset in the data path</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Prediction by the pre-trained deep neural network with phylogenetic tree weight regularizer.</p>
<p>prediction = deepbiome_predictoin(log, network_info, path_info, num_classes)</p>
<p>For LOOCV prediction, we can use this options.
prediction = deepbiome_predictoin(log, network_info, path_info, num_classes, number_of_fold=None, change_weight_for_each_fold=True)</p>
</dd></dl>

<dl class="function">
<dt id="deepbiome.deepbiome.deepbiome_get_trained_weight">
<code class="sig-prename descclassname">deepbiome.deepbiome.</code><code class="sig-name descname">deepbiome_get_trained_weight</code><span class="sig-paren">(</span><em class="sig-param">log, network_info, path_info, num_classes, weight_path, tree_level_list=['Genus', 'Family', 'Order', 'Class', 'Phylum'], verbose=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/deepbiome/deepbiome.html#deepbiome_get_trained_weight"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#deepbiome.deepbiome.deepbiome_get_trained_weight" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for prediction by the pretrained deep neural network with phylogenetic tree weight regularizer.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>log (logging instance) :</strong></dt><dd><p>python logging instance for logging</p>
</dd>
<dt><strong>network_info (dictionary) :</strong></dt><dd><p>python dictionary with network_information</p>
</dd>
<dt><strong>path_info (dictionary):</strong></dt><dd><p>python dictionary with path_information</p>
</dd>
<dt><strong>num_classes (int):</strong></dt><dd><p>number of classes for the network. 0 for regression, 1 for binary classificatin.</p>
</dd>
<dt><strong>weight_path (string):</strong></dt><dd><p>path of the model weight</p>
</dd>
<dt><strong>tree_level_list (list):</strong></dt><dd><p>name of each level of the given reference tree weights
default=[‘Genus’, ‘Family’, ‘Order’, ‘Class’, ‘Phylum’]</p>
</dd>
<dt><strong>verbose (boolean):</strong></dt><dd><p>show the log if True
default=True</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>list of pandas dataframe:</dt><dd><p>the trained model’s weight</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Trained weight of the deep neural network with phylogenetic tree weight regularizer.</p>
<p>tree_weight_list = deepbiome_get_trained_weight(log, network_info, path_info, num_classes, weight_path)</p>
</dd></dl>

<dl class="function">
<dt id="deepbiome.deepbiome.deepbiome_taxa_selection_performance">
<code class="sig-prename descclassname">deepbiome.deepbiome.</code><code class="sig-name descname">deepbiome_taxa_selection_performance</code><span class="sig-paren">(</span><em class="sig-param">log, network_info, path_info, num_classes, true_tree_weight_list, trained_weight_path_list, tree_level_list=['Genus', 'Family', 'Order', 'Class', 'Phylum'], lvl_category_dict=None, verbose=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/deepbiome/deepbiome.html#deepbiome_taxa_selection_performance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#deepbiome.deepbiome.deepbiome_taxa_selection_performance" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for prediction by the pretrained deep neural network with phylogenetic tree weight regularizer.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>log (logging instance) :</strong></dt><dd><p>python logging instance for logging</p>
</dd>
<dt><strong>network_info (dictionary) :</strong></dt><dd><p>python dictionary with network_information</p>
</dd>
<dt><strong>path_info (dictionary):</strong></dt><dd><p>python dictionary with path_information</p>
</dd>
<dt><strong>num_classes (int):</strong></dt><dd><p>number of classes for the network. 0 for regression, 1 for binary classificatin.</p>
</dd>
<dt><strong>true_tree_weight_list (ndarray):</strong></dt><dd><p>lists of the true weight information with the shape of (k folds, number of weights)
<cite>true_tree_weight_list[0][0]</cite> is the true weight information between the first and second layers for the first fold. It is a numpy array with the shape of (number of nodes for the first layer, number of nodes for the second layer).</p>
</dd>
<dt><strong>trained_weight_path_list (list):</strong></dt><dd><p>lists of the path of trained weight for each fold.</p>
</dd>
<dt><strong>tree_level_list (list):</strong></dt><dd><p>name of each level of the given reference tree weights
default=[‘Genus’, ‘Family’, ‘Order’, ‘Class’, ‘Phylum’]</p>
</dd>
<dt><strong>verbose (boolean):</strong></dt><dd><p>show the log if True
default=True</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>summary (numpy array):</dt><dd><p>summary of the taxa selection performance</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>The taxa selection performance of the trained deep neural network with phylogenetic tree weight regularizer.</p>
<p>summary = deepbiome_taxa_selection_performance(log, network_info, path_info, num_classes)</p>
</dd></dl>

<dl class="function">
<dt id="deepbiome.deepbiome.deepbiome_draw_phylogenetic_tree">
<code class="sig-prename descclassname">deepbiome.deepbiome.</code><code class="sig-name descname">deepbiome_draw_phylogenetic_tree</code><span class="sig-paren">(</span><em class="sig-param">log, network_info, path_info, num_classes, file_name='%%inline', img_w=500, branch_vertical_margin=20, arc_start=0, arc_span=360, node_name_on=True, name_fsize=10, tree_weight_on=True, tree_weight=None, tree_level_list=['Genus', 'Family', 'Order', 'Class', 'Phylum'], weight_opacity=0.4, weight_max_radios=10, phylum_background_color_on=True, phylum_color=[], phylum_color_legend=False, show_covariates=True, verbose=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/deepbiome/deepbiome.html#deepbiome_draw_phylogenetic_tree"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#deepbiome.deepbiome.deepbiome_draw_phylogenetic_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>Draw phylogenetic tree</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>log (logging instance) :</strong></dt><dd><p>python logging instance for logging</p>
</dd>
<dt><strong>network_info (dictionary) :</strong></dt><dd><p>python dictionary with network_information</p>
</dd>
<dt><strong>path_info (dictionary):</strong></dt><dd><p>python dictionary with path_information</p>
</dd>
<dt><strong>num_classes (int):</strong></dt><dd><p>number of classes for the network. 0 for regression, 1 for binary classificatin.</p>
</dd>
<dt><strong>file_name (str):</strong></dt><dd><p>name of the figure for save.
- “<em>.png”, “</em>.jpg”
- “%%inline” for notebook inline output.
default=”%%inline”</p>
</dd>
<dt><strong>img_w (int):</strong></dt><dd><p>image width (pt)
default=500</p>
</dd>
<dt><strong>branch_vertical_margin (int):</strong></dt><dd><p>vertical margin for branch
default=20</p>
</dd>
<dt><strong>arc_start (int):</strong></dt><dd><p>angle that arc start
default=0</p>
</dd>
<dt><strong>arc_span (int):</strong></dt><dd><p>total amount of angle for the arc span
default=360</p>
</dd>
<dt><strong>node_name_on (boolean):</strong></dt><dd><p>show the name of the last leaf node if True
default=False</p>
</dd>
<dt><strong>name_fsize (int):</strong></dt><dd><p>font size for the name of the last leaf node
default=10</p>
</dd>
<dt><strong>tree_weight_on (boolean):</strong></dt><dd><p>show the amount and the direction of the weight for each edge in the tree by circle size and color.
default=True</p>
</dd>
<dt><strong>tree_weight (ndarray):</strong></dt><dd><p>reference tree weights
default=None</p>
</dd>
<dt><strong>tree_level_list (list):</strong></dt><dd><p>name of each level of the given reference tree weights
default=[‘Genus’, ‘Family’, ‘Order’, ‘Class’, ‘Phylum’]</p>
</dd>
<dt><strong>weight_opacity  (float):</strong></dt><dd><p>opacity for weight circle
default= 0.4</p>
</dd>
<dt><strong>weight_max_radios (int):</strong></dt><dd><p>maximum radios for weight circle
default= 10</p>
</dd>
<dt><strong>phylum_background_color_on (boolean):</strong></dt><dd><p>show the background color for each phylum based on <cite>phylumn_color</cite>.
default= True</p>
</dd>
<dt><strong>phylum_color (list):</strong></dt><dd><p>specify the list of background colors for phylum level. If <cite>phylumn_color</cite> is empty, it will arbitrarily assign the color for each phylum.
default= []</p>
</dd>
<dt><strong>phylum_color_legend (boolean):</strong></dt><dd><p>show the legend for the background colors for phylum level
default= False</p>
</dd>
<dt><strong>show_covariates (boolean):</strong></dt><dd><p>show the effect of the covariates
default= True</p>
</dd>
<dt><strong>verbose (boolean):</strong></dt><dd><p>show the log if True
default=True</p>
</dd>
<dt><strong>Returns</strong></dt><dd></dd>
<dt><strong>——-</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Draw phylogenetic tree</p>
<p>deepbiome_draw_phylogenetic_tree(log, network_info, path_info, num_classes, file_name = “%%inline”)</p>
</dd></dl>

<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
</tbody>
</table>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="example_with_one_input_file.html" class="btn btn-neutral float-right" title="Example : k fold cross-validation with an input file" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="prerequisites.html" class="btn btn-neutral float-left" title="Prerequisites" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Youngwon Choi

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>